packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,162 @@
+      const deadPromises = new Set(deadPromisesRaw.split(','));
+      syscall.vatstoreDelete('deadPromises');
+      for (const [vpid, watches] of watchedPromiseTable.entries()) {
+        if (deadPromises.has(vpid)) {
+          watchedPromiseTable.delete(vpid);
+          for (const watch of watches) {
+            const [watcher, ...args] = watch;
+            if (watcher.onRejected) {
+              watcher.onRejected('vat upgraded', ...args);
+            }
+          }
+        } else {
+          const p = convertSlotToVal(vpid);
+          pseudoThen(p, vpid);
+        }
+      }
+    }
+  }
+
+  // defined promise watcher objects indexed by kindHandle
+  let promiseWatcherByKindTable;
+
+  function preparePromiseWatcherTables() {
+    let watcherTableID = syscall.vatstoreGet('watcherTableID');
+    if (watcherTableID) {
+      promiseWatcherByKindTable = convertSlotToVal(watcherTableID);
+    } else {
+      promiseWatcherByKindTable = makeScalarBigMapStore(
+        'promiseWatcherByKind',
+        { durable: true },
+      );
+      watcherTableID = convertValToSlot(promiseWatcherByKindTable);
+      syscall.vatstoreSet('watcherTableID', watcherTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watcherTableID);
+    }
+
+    let watchedPromiseTableID = syscall.vatstoreGet('watchedPromiseTableID');
+    if (watchedPromiseTableID) {
+      watchedPromiseTable = convertSlotToVal(watchedPromiseTableID);
+    } else {
+      watchedPromiseTable = makeScalarBigMapStore('watchedPromises', {
+        keySchema: M.string(),
+        durable: true,
+      });
+      watchedPromiseTableID = convertValToSlot(watchedPromiseTable);
+      syscall.vatstoreSet('watchedPromiseTableID', watchedPromiseTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watchedPromiseTableID);
+    }
+  }
+
+  function watchPromise(p, watcher, ...args) {
+    const watcherVref = convertValToSlot(watcher);
+    assert(watcherVref, 'invalid watcher');
+    const { virtual } = parseVatSlot(watcherVref);
+    assert(virtual, 'promise watcher must be a virtual object');
+    if (watcher.onFulfilled) {
+      assert.typeof(watcher.onFulfilled, 'function');
+    }
+    if (watcher.onRejected) {
+      assert.typeof(watcher.onRejected, 'function');
+    }
+
+    const vpid = convertValToSlot(p);
+    assert(vpid, 'invalid promise');
+    const { type } = parseVatSlot(vpid);
+    assert(type === 'promise', 'watchPromise only watches promises');
+    if (watchedPromiseTable.has(vpid)) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.set(vpid, harden([...watches, [watcher, ...args]]));
+    } else {
+      watchedPromiseTable.init(vpid, harden([[watcher, ...args]]));
+      pseudoThen(p, vpid);
+    }
+  }
+
+  function providePromiseWatcher(
+    kindHandle,
+    fulfillHandler = x => x,
+    rejectHandler = x => x,
```

Comment:
Default reject should throw
```suggestion
    rejectHandler = x => { throw x; },
```

I know we don't yet have a good mechanism to report unhandled rejections, but at least it would be the correct default behavior.

Reply:
In our long discussion we talked about this and concluded that throwing would be Bad.  I don't recall the reasoning however.  I don't have a particular opinion on this myself, other than a nagging worry about who would be catching this if it threw.

Reply:
From what I remember we mentioned we were not going to try and chain watchers for now (it's very much possible, but we need a proper virtual promise abstraction then), which meant that the user's watcher should handle errors within the watcher.

However I firmly believe we shouldn't explicitly swallow errors if we can help. I think we should root all watcher callbacks in a new call stack / turn, and let the errors bubble to whatever future unhandled rejection logic may exist.

Reply:
Brian and I just discussed this and persuaded ourselves that you are right about how the default reject handler should be written.  I think there's also an analogous tweak to do in the case where the user provides their own watcher facet (instead of using the thing handed back by `providePromiseWatcher`) and their watcher facet has no `onRejected` method.